compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -313,12 +315,122 @@ getTypeClassDictionary world = \case
+                Left _ -> pure e
+                Right dval -> pure (dvalBody dval)
+
+        ETmApp e1 e2 -> do
+            e1' <- go e1
+            case stripLoc e1' of
+                ETmLam (x,_) b -> do
+                    pure (applySubstInExpr (exprSubst x e2) b)
+                _ ->
+                    pure (ETmApp e1' e2)
+
+        ETyApp e t -> do
+            e' <- go e
+            case stripLoc e' of
+                ETyLam (x,_) b -> do
+                    pure (applySubstInExpr (typeSubst x t) b)
+                _ ->
+                    pure (ETyApp e' t)
+
+        ELocation l e -> ELocation l <$> go e
+
+        e -> pure e
+
+stripLoc :: Expr -> Expr
+stripLoc = \case
+    ELocation _ e -> stripLoc e
+    e -> e
+
+calcPartyLiterals :: Expr -> HasNoPartyLiterals
+calcPartyLiterals e = HasNoPartyLiterals (cata go e)
   where
-
-    go :: ExprF (Expr, Info) -> (Expr, Info)
     go = \case
+        EBuiltinF (BEParty _) -> False
+        f -> and f
+
+-- | Attempt to lift a closed expression to the top level. Returns either
+-- a variable expression that references the lifted expression, or
+-- returns the original expression.
+liftClosedExpr :: Int -> Expr -> Simplifier Expr
+liftClosedExpr d e = do
+    cache <- gets sCache
+    case Map.lookup e cache of
+        Just name -> do
+            EVal <$> selfQualify name
+
+        Nothing -> do
+            world <- gets sWorldExtended
+            version <- gets sVersion
+            case runGamma world version (typeOf' e) of
+                Right ty -> do
+                    name <- freshExprVarNameFor e
+                    e' <- inlineClosedExpr d world e
+                    addDefValue DefValue
+                        { dvalBinder = (name, ty)
+                        , dvalBody = e'
+                        , dvalLocation = Nothing
+                        , dvalNoPartyLiterals = calcPartyLiterals e'
+                        , dvalIsTest = IsTest False
+                        }
+                    EVal <$> selfQualify name
+
+                Left _ ->
+                    pure e
+
+simplifyExpr :: Expr -> Simplifier Expr
+simplifyExpr = simplifyExpr' 3
+
+simplifyExpr' :: Int -> Expr -> Simplifier Expr
+simplifyExpr' d e | d < 0 = pure e
+simplifyExpr' d e = fmap fst (cata go' e)
+  where
+    go' :: ExprF (Simplifier (Expr, Info)) -> Simplifier (Expr, Info)
+    go' ms = do
+        es <- sequence ms
+        world <- gets sWorldExtended
+        let v' = freeVarsStep (fmap (freeVars . snd) es)
+
+        -- We decide here whether it's worth performing constant lifting
+        -- for closed terms immediately under the current term. We want
+        -- to avoid creating unnecessary bindings, so we only perform
+        -- constant lifting when a closed term would become non-closed,
+        -- thereby grouping all the closed subterms together into a single
+        -- lift. If possible, we also want to lift above a lambda, even
```

Comment:
Don’t you mean that you want to lift below a lambda?

Reply:
ah, english is hard. I mean lift from below the lambda, to above the lambda (i.e. to the top level)